src/nativeops.jl

```diff
@@ -0,0 +1,336 @@
+    # for finding the correct operator.
+    p_is = p_lo = p_la = p_dbd = pointer_from_objref(op)
+
+    c_wrapper_fb = cfunction(_wrapper_fb, Bool, (Cint, Ptr{Ptr{Void}}, Ptr{Cint}, Ptr{Void}))
+    c_wrapper_infer = cfunction(_wrapper_infer, Bool, (Cint, Ptr{Cint}, Ptr{Ptr{Cuint}}, Ptr{Void}))
+    c_wrapper_list_outputs = cfunction(_wrapper_list_outputs, Bool, (Ptr{Ptr{Ptr{Cchar}}}, Ptr{Void}))
+    c_wrapper_list_arguments = cfunction(_wrapper_list_arguments, Bool, (Ptr{Ptr{Ptr{Cchar}}}, Ptr{Void}))
+    c_wrapper_declare_backward_dependency = cfunction(_wrapper_declare_backward_dependency, Bool, (Ptr{Cint}, Ptr{Cint}, Ptr{Cint}, Ptr{Cint}, Ptr{Ptr{Cint}}, Ptr{Void}))
+
+    # Setting up for handling backward/forward. Each function has a condition for a task
+    # to wait on and a libuv callback that notifies that condition, to handle the call.
+    cond_forward = Condition()
+    cond_backward = Condition()
+    cb_f = Base.SingleAsyncWork(data -> notify(cond_forward))
+    cb_b = Base.SingleAsyncWork(data -> notify(cond_backward))
+
+    # The return value of the callbacks is stored in _FB and the first element is the
+    # libuv handle.
+    r_forward = Ref(_FB(cb_f.handle))
+    r_backward = Ref(_FB(cb_f.handle))
+
+    p_f = Base.unsafe_convert(Ptr{Void}, r_forward)
+    p_b = Base.unsafe_convert(Ptr{Void}, r_backward)
+
+    # Task for handling forward
+    @schedule begin
+      try
+        while true
+           wait(cond_forward)
+           cond_forward = Condition()
+           _entry_forward(op, r_forward[])
+        end
+      catch
+        rethrow()
+      finally
+        Base.close(cb_f)
+      end
+    end
+
+    # Task for handling backward
+    @schedule begin
+      try
+        while true
+           wait(cond_backward)
+           cond_backward = Condition()
+           _entry_backward(op, r_backward[])
+        end
+      catch
+        rethrow()
+      finally
+        Base.close(cb_f)
+      end
+    end
+
+    new(c_wrapper_fb, c_wrapper_fb, c_wrapper_infer, c_wrapper_list_outputs,
+        c_wrapper_list_arguments,  c_wrapper_declare_backward_dependency,
+        p_f, p_b, p_is, p_lo, p_la, p_dbd)
+  end
+end
+
+###
+# Infer and list are called in sync.
+###
+function _wrapper_infer(size :: Cint, ndims :: Ptr{Cint}, tensor_shapes :: Ptr{Ptr{Cuint}}, _op :: Ptr{Void})
+  try
+    op = unsafe_pointer_to_objref(_op) :: Operator
+
+    n_in = length(list_arguments(op))
+    n_out = length(list_outputs(op))
+    @assert size == n_in + n_out
+
+    shapes = Vector{Cuint}[]
+    for i in 1:n_in
+      # Get size of array and create julia array.
+      jj = unsafe_load(ndims, i)
+      r_shapes = zeros(Cuint, jj)
+      # Get pointer to array
+      tshapes = unsafe_load(tensor_shapes, i)
+      # Copy values
+      for j in 1:jj
+        r_shapes[j] = unsafe_load(tshapes, j)
+      end
+      push!(shapes, r_shapes)
+    end
+
+    ishape, oshape = infer_shape(op, shapes)
```

Comment:
The shapes should be reversed before passing into `infer_shape` and then reversed before passing back to libmxnet. Because of row-majoring vs col-majoring.